In [1]:
import numpy as np
from scipy import stats

np.random.seed(5)

daily_temps = np.random.normal(
    loc=23,
    scale=2.5,
    size=30
)

x_bar = daily_temps.mean()
s = daily_temps.std(ddof=1)
n = len(daily_temps)

print("Варіант 5 — Дніпро")
print("Кількість спостережень:", n)
print("Вибіркове середнє:", x_bar)
print("Стандартне відхилення:", s)

Варіант 5 — Дніпро
Кількість спостережень: 30
Вибіркове середнє: 22.92771835022647
Стандартне відхилення: 2.4151177465764073


Для Дніпра було згенеровано вибірку з 30 поточних денних
температур липня.

Вибіркове середнє становить приблизно 22.93 °C,
а вибіркове стандартне відхилення — 2.42 °C.

Ці значення є точковими оцінками відповідних параметрів
генеральної сукупності.

Стара кліматична норма для Дніпра:

μ₀ = 21.6 °C.

Формулюємо гіпотези:

H₀: μ = 21.6 °C
Стара норма середньої липневої температури залишається
правильною.

H₁: μ ≠ 21.6 °C
Середня липнева температура відрізняється від старої норми.

Обираємо двобічний тест, оскільки перевіряємо, чи стара
норма взагалі перестала відповідати дійсності. Нас цікавить
як можливе підвищення, так і можливе зниження температури.

Тип тесту обрано до обчислення p-value.

In [3]:
mu0 = 21.6

se = s / np.sqrt(n)

z = (x_bar - mu0) / se

p_value = 2 * (
    1 - stats.norm.cdf(abs(z))
)

print("Стара норма μ₀:", mu0)
print("Середнє x̄:", x_bar)
print("SE:", se)
print("z:", z)
print("p-value:", p_value)

Стара норма μ₀: 21.6
Середнє x̄: 22.92771835022647
SE: 0.44093815627698113
z: 3.011121472083367
p-value: 0.002602847038782663


Стандартна похибка становить приблизно 0.441 °C.

z-статистика дорівнює приблизно 3.01. Це означає,
що вибіркове середнє 22.93 °C знаходиться приблизно
на 3.01 стандартної похибки вище від старої норми
21.6 °C.

Для двобічного тесту отримано:

p-value ≈ 0.0026.

Це означає, що якби H₀ була правильною і справжня
середня температура дорівнювала 21.6 °C, ймовірність
отримати вибіркове середнє з таким або ще більшим
відхиленням в будь-який бік становила б приблизно 0.26%.

p-value не є ймовірністю того, що H₀ правильна.

In [4]:
alpha = 0.05

if p_value < alpha:
    print("Відхиляємо H₀")
else:
    print("Не відхиляємо H₀")

Відхиляємо H₀


При рівні значущості α = 0.05 отримано:

p-value ≈ 0.0026 < 0.05.

Тому нульову гіпотезу H₀ відхиляємо.

Наявні дані дають статистично значущі підстави вважати,
що середня липнева температура Дніпра відрізняється
від старої кліматичної норми 21.6 °C
(p ≈ 0.0026, двобічний тест, α = 0.05).

У нашій вибірці середнє становить приблизно 22.93 °C,
тобто воно вище за стару норму.

In [5]:
ci_lower = x_bar - 1.96 * se
ci_upper = x_bar + 1.96 * se

print("95% довірчий інтервал:")
print(ci_lower, ci_upper)

print(
    "Чи входить μ₀ в інтервал?",
    ci_lower <= mu0 <= ci_upper
)

95% довірчий інтервал:
22.063479563923586 23.791957136529355
Чи входить μ₀ в інтервал? False


95% довірчий інтервал для середньої липневої температури:

[22.06 °C; 23.79 °C].

Стара норма μ₀ = 21.6 °C не входить до цього інтервалу.

Отже, результат довірчого інтервалу узгоджується
з перевіркою гіпотези: при α = 0.05 ми відхилили H₀.

Помилка I роду:

Ми відхиляємо H₀ і робимо висновок, що стара кліматична
норма 21.6 °C для Дніпра вже не відповідає дійсності,
хоча насправді вона все ще є правильною.

Помилка II роду:

Ми не відхиляємо H₀ і вважаємо, що стара норма 21.6 °C
ще відповідає дійсності, хоча насправді середня липнева
температура Дніпра вже змінилася.

Для кліматологів, які готують офіційний довідник, обидві
помилки небажані. Помилка I роду могла б призвести до
оновлення офіційної норми без достатніх підстав, тому перед
офіційною зміною довідника важливо мати переконливі дані
та достатню кількість спостережень.

Контрольні питання

1. Що означає отримане p-value?

Для Дніпра отримано p-value ≈ 0.0026.

Це означає, що якби нульова гіпотеза була правильною
і справжня середня липнева температура дорівнювала
21.6 °C, то ймовірність отримати результат із таким
або ще більшим відхиленням у будь-який бік становила б
приблизно 0.26%.

p-value не показує ймовірність того, що H₀ правильна.
Воно обчислюється за припущення, що H₀ вже є істинною.


2. Чому тип тесту потрібно обирати до результату?

Тип тесту потрібно визначити на основі дослідницького
питання до того, як ми побачимо результат вибірки.

Якщо вибрати однобічний або двобічний тест після перегляду
даних лише тому, що один із них дає менше p-value,
це збільшує ризик отримати статистично значущий результат
внаслідок підлаштування аналізу під дані.

У цій роботі був заздалегідь обраний двобічний тест,
оскільки нас цікавило будь-яке відхилення від 21.6 °C.


3. Що означають помилки I та II роду?

Помилка I роду означала б зробити висновок, що стара
кліматична норма Дніпра 21.6 °C вже неправильна,
коли насправді вона все ще правильна.

Помилка II роду означала б не виявити реальної зміни:
залишити стару норму 21.6 °C, хоча справжня середня
липнева температура вже змінилася.

Це різні помилки: у першому випадку ми помилково
відхиляємо правильну H₀, а в другому — не відхиляємо
H₀, яка насправді неправильна.


4. Чи узгоджується тест із довірчим інтервалом?

Так.

У тесті отримано p ≈ 0.0026 < 0.05, тому H₀ про
μ = 21.6 °C була відхилена.

95% довірчий інтервал становить приблизно
[22.06 °C; 23.79 °C], і значення 21.6 °C до нього
не входить.

Результати узгоджуються, оскільки двобічний тест
при α = 0.05 пов'язаний із відповідним 95% довірчим
інтервалом: значення μ₀, які лежать за межами такого
інтервалу, відхиляються двобічним тестом на рівні 0.05.